x/staking/keeper/msg_server.go

```diff
@@ -247,35 +257,74 @@ func (k msgServer) Delegate(goCtx context.Context, msg *types.MsgDelegate) (*typ
 // BeginRedelegate defines a method for performing a redelegation of coins from a delegator and source validator to a destination validator
 func (k msgServer) BeginRedelegate(goCtx context.Context, msg *types.MsgBeginRedelegate) (*types.MsgBeginRedelegateResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
+
 	valSrcAddr, err := sdk.ValAddressFromBech32(msg.ValidatorSrcAddress)
 	if err != nil {
 		return nil, err
 	}
+	valDstAddr, err := sdk.ValAddressFromBech32(msg.ValidatorDstAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	srcValidator, found := k.GetValidator(ctx, valSrcAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+	dstValidator, found := k.GetValidator(ctx, valDstAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
 	delegatorAddress, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
 	if err != nil {
 		return nil, err
 	}
-	shares, err := k.ValidateUnbondAmount(
-		ctx, delegatorAddress, valSrcAddr, msg.Amount.Amount,
-	)
+
+	delegation, found := k.GetDelegation(ctx, delegatorAddress, valSrcAddr)
+	if !found {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"delegation with delegator %s not found for validator %s",
+			msg.DelegatorAddress, msg.ValidatorSrcAddress,
+		)
+	}
+
+	srcShares, err := k.ValidateUnbondAmount(ctx, delegatorAddress, valSrcAddr, msg.Amount.Amount)
+	if err != nil {
+		return nil, err
+	}
+	dstShares, err := dstValidator.SharesFromTokensTruncated(msg.Amount.Amount)
 	if err != nil {
 		return nil, err
 	}
 
+	// if this is a validator self-bond, the new liquid delegation cannot fall below the self-bond * bond factor
+	if delegation.ValidatorBond {
+		if err := k.SafelyDecreaseValidatorBond(ctx, &srcValidator, srcShares); err != nil {
+			return nil, err
+		}
+	}
+
+	// If this delegation from a liquid staker, the delegation on the new validator
+	// cannot exceed that validator's self-bond cap
+	// The liquid shares from the source validator should get moved to the destination validator
+	if k.DelegatorIsLiquidStaker(delegatorAddress) {
+		if err := k.SafelyIncreaseValidatorLiquidShares(ctx, &dstValidator, dstShares); err != nil {
+			return nil, err
+		}
+		k.DecreaseValidatorLiquidShares(ctx, &srcValidator, srcShares)
```

Comment:
Should check for `err` from `k.DecreaseValidatorTotalLiquidShares`, it could return `ErrValidatorLiquidSharesUnderflow`

Credit for catching goes to @hieuvubk for spotting

Reply:
good catch! resolved [here](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/73e72b9ef30f2315adeeefa33aa2526cf23ba8b7)